### PR TITLE
Feature/xp rpc

### PR DESCRIPTION
**Pull request by scottmunday84**
gRPC added, HookHorror basis built, and the xp_rpc plugin base developed.

### CODE REVIEW

File: src/controller/nwnx.ini [hunk 1]
```diff
@@ -84,3 +84,9 @@ gracefulShutdownMessage = Server is shutting down NOW!
 # Value = seconds to wait before actual shutdown
 # default: 5 seconds
 gracefulShutdownMessageWait = 5
+
+#logServiceUrl=
```
Comment (CromFr) [inline src/controller/nwnx.ini:88]: Some documentation for these values would be nice :)

File: src/plugins/xp_rpc/xp_rpc.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+clients:
```
Comment (CromFr) [inline src/plugins/xp_rpc/xp_rpc.yml:1]: This file would need some documentation

**Review by CromFr — COMMENTED** [review]
First pass, I haven't looked closely at the code yet

File: .gitignore [hunk 3]
```diff
@@ -23,6 +23,7 @@
 /src/plugins/*/Release/
 
 .idea/
+vcpkg/
```
[the remark this answers is not in the document]
Comment (scottmunday84) [reply .gitignore:26]: This is fine, but does this mean that any updates to the packages will also have to installed into this "other" repository? I can do that, but will have to build a PR on that submodule repo too.

File: grpc-build.ps1 [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# HookHorror
```
[the remark this answers is not in the document]
Comment (scottmunday84) [reply grpc-build.ps1:1]: Wasn't aware of generators. Very cool. I'll add the setup to the branch.

File: meson-build-watcher.ps1 [hunk 5]
```diff
@@ -0,0 +1 @@
+%comspec% /c """C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\Common7\Tools\VsDevCmd.bat"" & ninja reconfigure"
```
[the remark this answers is not in the document]
Comment (scottmunday84) [reply meson-build-watcher.ps1:1]: Not necessarily, but I use that for my JetBrains watchers, so it auto updates the compilation DB.

File: src/HookHorror/proto/log.grpc.pb.cc [hunk 6]
```diff
@@ -0,0 +1,302 @@
+// Generated by the gRPC C++ plugin.
```
[the remark this answers is not in the document]
Comment (scottmunday84) [reply src/HookHorror/proto/log.grpc.pb.cc:1]: Will do.

File: src/plugins/xp_rpc/RpcPlugin.h [hunk 7]
```diff
@@ -0,0 +1,59 @@
+/***************************************************************************
+    NWNX FastBoot - NWN2Server Zero Module Copy (Startup Time Improvements)
```
[the remark this answers is not in the document]
Comment (scottmunday84) [reply src/plugins/xp_rpc/RpcPlugin.h:2]: Yeah. Clearly copied the basis of the plugin. :P

File: src/HookHorror/LogClient.cpp [hunk 8]
```diff
@@ -0,0 +1,24 @@
+//
```
[the remark this answers is not in the document]
Comment (scottmunday84) [reply src/HookHorror/LogClient.cpp:1]: rpc all lowercase is fine. But HookHorror is my codename for the work. Renaming it to rpc, less of me is in this project.

File: src/controller/nwnx.ini (hunk 1 above)
Comment (scottmunday84) [reply src/controller/nwnx.ini:88]: Will do.

File: src/plugins/xp_rpc/xp_rpc.yml (hunk 2 above)
Comment (scottmunday84) [reply src/plugins/xp_rpc/xp_rpc.yml:1]: Yes, I will add documentation for this. It is basically a YAML file that contains a client list of unsecure names and routes. YAML is much easier than INI files, especially with building out a map of client names to routes. Behind the scenes, the xp_rpc plugin will use the client name and route to that location based on the NWNX* function called.

For example, all NWNX* methods require a plugin name, a function name, and a param1. In this plugin, the plugin name is "RPC", the function name is the "client" map key (i.e. "test", "crafting", "health", etc.) and the param1 will be typically used internal to the client to distinguish which function needs to be called. Also could be built as a route or whatever you need out of it.

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (scottmunday84) [reply .gitignore]: Done.

File: grpc-build.ps1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (scottmunday84) [reply grpc-build.ps1]: This is fixed; however, the grpc_files (output) don't have the full paths as expected. Manually added protobuf files to controller, gui, and xp_rpc projects.

File: meson-build-watcher.ps1 (hunk 5 above)
[the remark this answers is not in the document]
Comment (scottmunday84) [reply meson-build-watcher.ps1]: Moved to a folder called "dev". Ignore the entire folder.

**Comment by scottmunday84** [comment]
All comments have been addressed. Rebuilt, works well.

**Comment by CromFr** [comment]
Wow, the CI is taking more than 1hour to build all the vcpkg dependencies !

It's not dramatic because it should happen only once because the CI results are cached, however you'll need to add this to the build.yml (once rebased on the main branch):

```diff
      # Install vcpkg dependencies
      - name: ↳ Install & build dependencies
        if: steps.vcpkg_cache.outputs.cache-hit != 'true'
+       timeout-minutes: 120
        shell: bash
        run: |
          vcpkg/vcpkg.exe install --triplet=x86-windows-static-md --disable-metrics
```

**Comment by scottmunday84** [comment]
> Wow, the CI is taking more than 1hour to build all the vcpkg dependencies !
> 
> It's not dramatic because it should happen only once because the CI results are cached, however you'll need to add this to the build.yml (once rebased on the main branch):
> 
> ```diff
>       # Install vcpkg dependencies
>       - name: ↳ Install & build dependencies
>         if: steps.vcpkg_cache.outputs.cache-hit != 'true'
> +       timeout-minutes: 120
>         shell: bash
>         run: |
>           vcpkg/vcpkg.exe install --triplet=x86-windows-static-md --disable-metrics
> ```

Yeah, I was wondering (now that we have CI tasks). gRPC is a large set of tools and dependencies. I also added yaml-cpp (for my YAML processing), but it isn't nearly as large.

**Review by CromFr — CHANGES_REQUESTED** [review]

File: src/controller/controller.cpp [hunk 9]
```diff
@@ -23,6 +23,8 @@
 
 extern LogNWNX* logger;
 
+std::future<void> logServiceConnection;
```
[the remark this answers is not in the document]
Comment (scottmunday84) [reply src/controller/controller.cpp]: Moved to header file.

File: src/controller/controller.cpp [hunk 10]
```diff
@@ -73,6 +75,14 @@ NWNXController::NWNXController(SimpleIniConfig* config)
 	}
 	logger->Trace("NWN2 install dir: %s", nwninstalldir.c_str());
 	logger->Trace("NWN2 parameters: %s", parameters.c_str());
+
+	// Create log service (if necessary).
+	try {
+        logService = new NWNX4::RPC::Log::LogService(nwninstalldir.c_str());
```
[the remark this answers is not in the document]
Comment (scottmunday84) [reply src/controller/controller.cpp]: Set to current dir "." relative path.

File: src/controller/controller.h [hunk 11]
```diff
@@ -75,8 +77,9 @@ class NWNXController
 	std::string nwninstalldir;
 	bool initialized;
 	bool shuttingDown;
+    NWNX4::RPC::Log::LogService* logService;
```
[the remark this answers is not in the document]
Comment (scottmunday84) [reply src/controller/controller.h]: Now defaults to nullptr.

**Review by CromFr — APPROVED** [review]
LGTM !

**Pull request merged**